src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -307,6 +307,204 @@ void ReplicatedWriteLog<I>::arm_periodic_stats() {
+      sync_point_entries[pmem_entry->sync_gen_number] = sync_point_entry;
+      missing_sync_points.erase(pmem_entry->sync_gen_number);
+      m_current_sync_gen = pmem_entry->sync_gen_number;
+    } else if (pmem_entry->is_write()) {
+      ldout(m_image_ctx.cct, 20) << "Entry " << entry_index
+                                 << " is a write. pmem_entry=[" << *pmem_entry << "]" << dendl;
+      auto write_entry =
+        std::make_shared<WriteLogEntry>(nullptr, pmem_entry->image_offset_bytes, pmem_entry->write_bytes);
+      write_entry->pmem_buffer = D_RW(pmem_entry->write_data);
+      log_entry = write_entry;
+    } else if (pmem_entry->is_writesame()) {
+      ldout(m_image_ctx.cct, 20) << "Entry " << entry_index
+                                 << " is a write same. pmem_entry=[" << *pmem_entry << "]" << dendl;
+      auto ws_entry =
+        std::make_shared<WriteSameLogEntry>(nullptr, pmem_entry->image_offset_bytes,
+                                            pmem_entry->write_bytes, pmem_entry->ws_datalen);
+      ws_entry->pmem_buffer = D_RW(pmem_entry->write_data);
+      log_entry = ws_entry;
+    } else if (pmem_entry->is_discard()) {
+      ldout(m_image_ctx.cct, 20) << "Entry " << entry_index
+                                 << " is a discard. pmem_entry=[" << *pmem_entry << "]" << dendl;
+      auto discard_entry =
+        std::make_shared<DiscardLogEntry>(nullptr, pmem_entry->image_offset_bytes, pmem_entry->write_bytes,
+                                          m_discard_granularity_bytes);
+      log_entry = discard_entry;
+    } else {
+      lderr(m_image_ctx.cct) << "Unexpected entry type in entry " << entry_index
+                             << ", pmem_entry=[" << *pmem_entry << "]" << dendl;
+      ceph_assert(false);
+    }
+
+    if (writer) {
+      ldout(m_image_ctx.cct, 20) << "Entry " << entry_index
+                                 << " writes. pmem_entry=[" << *pmem_entry << "]" << dendl;
+      if (!sync_point_entries[pmem_entry->sync_gen_number]) {
+        missing_sync_points[pmem_entry->sync_gen_number] = true;
+      }
+    }
+
+    log_entry->ram_entry = *pmem_entry;
+    log_entry->pmem_entry = pmem_entry;
+    log_entry->log_entry_index = entry_index;
+    log_entry->completed = true;
+
+    m_log_entries.push_back(log_entry);
+
+    entry_index = (entry_index + 1) % m_total_log_entries;
+  }
+
+  /* Create missing sync points. These must not be appended until the
+   * entry reload is complete and the write map is up to
+   * date. Currently this is handled by the deferred contexts object
+   * passed to new_sync_point(). These contexts won't be completed
+   * until this function returns.  */
+  for (auto &kv : missing_sync_points) {
+    ldout(m_image_ctx.cct, 5) << "Adding sync point " << kv.first << dendl;
+    if (0 == m_current_sync_gen) {
+      /* The unlikely case where the log contains writing entries, but no sync
+       * points (e.g. because they were all retired) */
```

Comment:
@dillaman Hi Jason, if some sync point retired, its associated writes retired. Here is the case that some writes are just on cache but their associated sync point isn't on cache yet. For example, [write1, syncpont1, write2] on the cache, then write1 and syncpont1 are retired, now only [write2] on the cache and write2's sync point has not been on the cache yet.

Reply:
Writes refer to a sync point that follows them in the log. We append writes as soon as possible so (in persist-on-flush mode) the aio_write can be completed to the caller as soon as we're done copying from their buffer. We don't append the sync point until we have to (aio_flush is called, or we hit our configured max writes per sync point, or we're inserting sync points after every append batch in persist-on-write mode). So if we crash before we persist the sync point for the current sync gen number, the last batch of writes will refer to a sync point that doesn't appear in the log.

There can be multiple sync points missing if there are multiple aio_flush calls (or inserted sync points) in progress at once. That happens because we don't make writes wait for prior writes (unless they overlap) or flushes. 

When writes exit the block guard, we associate them with the current sync gen number. If we think we're still on the caller's thread, we attempt to persist the user's write payload while we're still on that thread. This is just a memcpy (and an RDMA write if we're replicating). Since the PMDK persist calls are currently synchronous we do this in the caller's thread so they can all occur concurrently in HW (otherwise they'd occur one at a time on a WQ thread, leaving a lot of idle memory channel and RDMA fabric link-nanoseconds). If there was an async persist (/replicate) call, or HW data mover available to user mode libraries, we'd use that here. Once the payload data is persisted, the write is eligible for append.

Some writes take longer than others to persist, either because they're larger, the memcpy causes a page fault, or the persisting thread gets preempted. Writes can be passed by later writes and aio_flush calls when this happens (or just because they're called from different cores and aren't really happening in the same global order the application thinks they are).

Log entries and sync points are scheduled for append by placing them in the append list. That' a list of entries not yet persisted in the log, but in the order they will be persisted. Sync points must appear after all the entries that refer to them, so when the current sync gen number changes the current sync point waits on a Gather for all its writes to be scheduled for append. When those writes enter the append list, their sync point can enter the append list. Those writes won't be appended until their buffers persist, so a sync point is waiting for all of that before it enters the append list. That Gather also includes the previous sync point, so sync points always enter the append list in order.

Writes arriving while their prior sync point is still waiting to enter the append list can complete, enter the append list, and get appended before that prior sync point is appended. At this point if we crash, we have write log entries referring to two consecutive sync points that do not appear in the log.

Writes aren't eligible for flush until their sync point is (appended and) persisted (/replicated). So it might look like we could just drop these writes that refer to missing sync points. Actually we can't because writes are eligible for read as soon as they are appended. So while these writes can't have been flushed, they may have been read. An application that read this write may have taken action on its contents (e.g. sent you your Amazon order), so we don't allow that data to change by dropping these writes.

It's now been years since I wrote that stuff and worked on it every day, so if I've misremembered something here but this is the basic idea of the RWL write path here, and how that can look on restore.